# One-Sample Z-Test for Proportions: pandas-dev/pandas Statistical Audit

### Research Questions:
* **RQ1:** Apakah proporsi keberhasilan merge Pull Request (PR) secara signifikan berbeda dari standar klaim historis sebesar 60%? *(Fokus Analisis Member D)*
* **RQ2:** Berapa rentang estimasi probabilitas PR berhasil di-merge pada tingkat kepercayaan 95%? *(Dianalisis oleh Member C)*
* **RQ3:** Apakah rata-rata issue mingguan berubah secara signifikan setelah adanya major release? *(Dianalisis pada modul kelompok lainnya)*

**Member:** Muhammad Yunus Setiaji — Hypothesis Analyst (Member D)

---

## AI Usage Disclosure

**Member:** Muhammad Yunus Setiaji — Hypothesis Analyst | **Tools used:** Gemini

| Task | Tool | Prompt summary | Output modified? |
| :--- | :--- | :--- | :--- |
| Coding boilerplate untuk pemanggilan fungsi Z-Test | Gemini | "Meminta kerangka kode Python untuk memanggil fungsi z_test_one_sample dari modul lokal" | Ya — Variabel disesuaikan dengan data riil n dan x_bar dari repositori Pandas. |
| Pembuatan struktur template (scaffolding) laporan | Gemini | "Meminta struktur kosong 6-step procedure uji hipotesis" | Ya — Mengisi seluruh teks narasi langkah dan kesimpulan secara mandiri tanpa AI. |

**Written entirely without AI:** Teks pernyataan Hipotesis H0 dan Ha pada Step 1, analisis asumsi data pada Step 2, penulisan narasi keputusan statistik pada Step 5, serta seluruh paragraf evaluasi engineering pada Step 6 dan Analisis Integratif.

---

## Analisis Inferensia: Pengujian Hipotesis (Z-Test)

In [1]:
import sys
sys.path.append("../")
import numpy as np
from src.hypothesis import z_test_one_sample

# Data hulu yang diserahterimakan dari Member B dan Member C
n = 4042
x_bar = 0.62098  # Nilai point estimate proporsi sukses merge PR

# Standard deviasi untuk distribusi proporsi (Bernoulli)
sigma = np.sqrt(x_bar * (1 - x_bar))

# Nilai komparasi hipotesis awal (Klaim historis 60%)
mu0 = 0.60000 

# Eksekusi fungsi pengujian hipotesis dari src/hypothesis.py (Sesuai Tsun 2020 p. 306)
hasil = z_test_one_sample(
    x_bar=x_bar, 
    mu0=mu0, 
    sigma=sigma, 
    n=n, 
    alternative='two-sided', 
    alpha=0.05
)

print("=== LIVE REPORT CHECKPOINT ===")
print("Z-Statistic :", round(hasil["z_stat"], 4))
print("P-Value     :", round(hasil["p_value"], 4))
print("Keputusan   :", hasil["decision"])
print("Narasi      :", hasil["interpretation"])

=== LIVE REPORT CHECKPOINT ===
Z-Statistic : 2.7494
P-Value     : 0.006
Keputusan   : Reject H0
Narasi      : With a p-value of 0.0060 < 0.05, we have sufficient evidence to reject H0 at the 0.05 significance level.


### Prosedur Uji Hipotesis (6-Step Procedure) - Member D

**Step 1: State Hypotheses**
* $H_0: \mu = 0.60$ (Proporsi dari total pull request yang berhasil di-merge pada repositori Pandas adalah sama dengan angka historis 60%)
* $H_a: \mu \neq 0.60$ (Proporsi dari total pull request yang berhasil di-merge pada repositori Pandas secara signifikan tidak sama dengan angka historis 60%)

**Step 2: Assumptions Check**
* Data pull request yang digunakan bersifat independen karena ditarik langsung dari log aktivitas sistem Git repositori `pandas-dev/pandas`.
* Ukuran sampel data yang kita miliki sangat besar yaitu $n = 4042$, di mana nilai $n \cdot p \ge 10$ dan $n \cdot (1-p) \ge 10$ terpenuhi secara mutlak. Berdasarkan Teorema Limit Pusat (Central Limit Theorem), bentuk distribusi sampling dari proporsi ini akan berkonvergensi mendekati distribusi normal standar. Oleh karena itu, penggunaan uji satu sampel Z-Test ini dinilai valid.

**Step 3: Test Statistic Calculation**
* Formula konseptual komputasi merujuk pada Tsun (2020) halaman 306:
  $$Z = \frac{\bar{x} - \mu_0}{\sigma / \sqrt{n}}$$
* Melalui eksekusi fungsi `z_test_one_sample` pada modul `src/hypothesis.py`, diperoleh nilai Z-statistic sebesar **2.7494**.

**Step 4: P-value and Critical Region**
* Pengujian ini menggunakan tingkat signifikansi $\alpha = 0.05$ dengan tipe dua arah (two-sided).
* Wilayah kritis untuk menolak $H_0$ ditentukan pada batas nilai $Z < -1.96$ atau $Z > 1.96$.

**Step 5: Statistical Decision**
* Berdasarkan perhitungan komputasi dengan `scipy.stats`, diperoleh nilai p-value sebesar **0.0060**.
* Karena nilai p-value ($0.0060$) < $\alpha$ ($0.05$) dan nilai $Z$-stat ($2.7494$) berada di dalam wilayah kritis ($Z > 1.96$), maka keputusan statistik yang diambil secara tegas adalah **Reject $H_0$** (Menolak $H_0$).

**Step 6: Conclusion & Interpretation**
* Dengan tingkat signifikansi 5%, terdapat bukti statistik yang sangat kuat untuk menolak $H_0$. Hal ini menyimpulkan bahwa proporsi pull request yang berhasil masuk ke repositori Pandas berbeda secara nyata dari klaim standar 60%. Performa riil yang berada di angka 62.09% menunjukkan tingkat penerimaan kode yang lebih tinggi. Bagi maintainer, ini adalah sinyal positif bahwa alur kerja *code review* mereka berjalan produktif dan kontribusi komunitas tersaring dengan baik tanpa penumpukan *backlog*.

---

### Hubungan Integrasi Across Layers (Member D & Member C)
Hasil penolakan $H_0$ pada analisis **RQ1** ini terbukti sinkron secara matematis dengan hasil perhitungan interval pada **RQ2** yang digarap oleh Member C (Daffa). Pada lembar kerja sebelumnya, Daffa menemukan bahwa rentang 95% Confidence Interval berada di kisaran angka 0.6060 hingga 0.6359. 

Karena nilai parameter yang diuji pada $H_0$ ($\mu_0 = 0.6000$) berada di luar rentang interval kepercayaan tersebut (di bawah batas bawah 0.6060), maka secara otomatis hipotesis nol tersebut wajib ditolak pada tingkat signifikansi 5%. Keselarasan ini membuktikan validitas pipeline data kelompok kami dari awal hingga akhir.